### PR TITLE
[wip]generate accessed files list for workload

### PR DESCRIPTION
**Pull request by sctb512**
Generate an in-order accessed files list for workload. The accessed files list can be set as `prefetch-patterns` when converting OCI images to nydus format images.
Reference to https://github.com/containerd/stargz-snapshotter/tree/main/analyzer

`oerdctl` is a `nerdctl`-compatible cli to adapt the `nerdctl run` command for recording accessed files when task running.

###  basic usage
We can use `oerdctl run` just like `nerdctl run`. You don't need time to learn the command parameters, because they are the same as `nerdctl run`. When running a task, the accessed files list will be stored at the default disk path `accessed_files.txt`.
More information about `nerdctl run`: [nerdctl#whale-blue_square-nerdctl-run](https://github.com/containerd/nerdctl#whale-blue_square-nerdctl-run)
```shel
oerdctl run --rm --net host -it golang
```
###  special parameters
We can specify recording time with the `--wait-time` option.

```shell=
oerdctl run --rm --net host --wait-time 30 nginx
```

Otherwise, we can specify a string line to wait for with `--wait-line`, just like:

```shell=
oerdctl run --rm --net host --wait-line "ready for start up" nginx
```
The `oerdctl run` will kill task when getting expect string line from stand output.

We can also specify `oerdctl run` to wait for a signal with the `--wait-signal` option. Just like:

```shell=
oerdctl run --rm --net host --wait-signal nginx
```
The `oerdctl run` will kill task when getting `SIGINT` (also know as `ctrl + C`) signal.

The other option `--output` is also available to specify the output path for storing accessed files.

This PR is related to https://github.com/dragonflyoss/image-service/issues/807

Signed-off-by: Bin Tang <tangbin.bin@bytedance.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,95 +5,174 @@ go 1.18
...
+	github.com/google/go-cmp v0.5.9 // indirect
 	github.com/google/gofuzz v1.2.0 // indirect
 	github.com/google/uuid v1.2.0 // indirect
-	github.com/googleapis/gnostic v0.5.5 // indirect
-	github.com/gorilla/mux v1.8.0 // indirect
+	github.com/hashicorp/errwrap v1.1.0 // indirect
 	github.com/hashicorp/go-cleanhttp v0.5.1 // indirect
+	github.com/hashicorp/golang-lru v0.5.4 // indirect
 	github.com/imdario/mergo v0.3.12 // indirect
+	github.com/ipfs/bbloom v0.0.4 // indirect
```
Comment (imeoer) [inline go.mod:85]: Why need IPFS related packages?

File: cmd/optimize/main.go [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+	NamespaceEnvVar = "CONTAINERD_NAMESPACE"
+	// Default is the name of the default namespace
+	DefaultNamespace = "default"
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Flags = containerdFlags
+	app.Name = "optimize"
+	app.Usage = "Generate accessed files list for OCI format image."
```
Comment (imeoer) [inline cmd/optimize/main.go:57]: Thanks for adding this tool, maybe give some docs and mark the subcommand `experimental`?

File: go.mod [hunk 3]
```diff
@@ -5,95 +5,174 @@ go 1.18
...
+	github.com/ipfs/interface-go-ipfs-core v0.7.0 // indirect
+	github.com/ipld/go-codec-dagpb v1.3.2 // indirect
+	github.com/ipld/go-ipld-prime v0.11.0 // indirect
+	github.com/jbenet/goprocess v0.1.4 // indirect
+	github.com/josharian/intern v1.0.0 // indirect
 	github.com/json-iterator/go v1.1.12 // indirect
-	github.com/klauspost/compress v1.15.1 // indirect
+	github.com/klauspost/compress v1.15.11 // indirect
+	github.com/klauspost/cpuid/v2 v2.0.9 // indirect
+	github.com/libp2p/go-buffer-pool v0.0.2 // indirect
```
Comment (imeoer) [inline go.mod:115]: Is this necessary?

Comment (sctb512) [reply go.mod:115]: Maybe not, i am making a new design for adapting `nerdctl`.

**Comment by sctb512** [comment]
> Is it necessary to implement a `nerdctl`-like cli? Can we contribute the `oerdctl` codes to `nerdctl`? Or at least we should move the code to somewhere else repo.

Maybe not. It will copy many files from `nerdctl`. I will try to contribute the code to `nerdctl` first.

**Comment by AkihiroSuda** [comment]
I think this should be implemented as an OCI hook

https://github.com/containerd/nerdctl/pull/1686#issuecomment-1365605589

**Comment by sctb512** [comment]
> I think this should be implemented as an OCI hook
> 
> [containerd/nerdctl#1686 (comment)](https://github.com/containerd/nerdctl/pull/1686#issuecomment-1365605589)

Thanks for your reply. It is more acceptable if we modify fewer codes. I will try to implement it with OCI hook.

**Comment by sctb512** [comment]
We are implementing this features with OCI Hooks in another PR. It's time to close this.

**Pull request closed without merging**